CPFP

ptschip opened the pull request.

gandrewstone commented on src/globals.cpp:
```diff
@@ -274,6 +275,9 @@ CTweakRef<bool> displayArchInSubver("net.displayArchInSubver",
     "Show box architecture, 32/64bit, in node user agent string (subver)",
     &fDisplayArchInSubver);
 
+CTweak<bool> miningCPFP("mining.cpfp",
```
let's spell out mining.cpfp as mining.childPaysForParent, but variable name miningCPFP is fine.

ptschip replied:
done

gandrewstone commented on src/miner.cpp:
```diff
@@ -30,10 +30,16 @@
 #include <queue>
 #include <thread>
 
+// Track timing information for Score and Pacakge mining.
```
spelling: Package

Also, I'm just going from top to bottom here so may be missing something but shouldn't these be members of BlockAssembler?

ptschip replied:
fixed spelling

as for being part of block assembler, i don't think it makes
sense...these are global vars used to track peformance in the
miner_tests.cpp peformancetest section.

gandrewstone commented on src/miner.h:
```diff
@@ -22,20 +25,50 @@ class CWallet;
+    bool operator()(const CTxMemPool::txiter &a, const CTxMemPool::txiter &b) const { return &(*a) < &(*b); }
+};
+
+// A comparator that sorts transactions based on number of ancestors.
```
use doxygen compatible comment format /**

ptschip replied:
done

gandrewstone commented on src/miner.cpp:
```diff
@@ -473,6 +515,141 @@ void BlockAssembler::addScoreTxs(std::vector<const CTxMemPoolEntry *> *vtxe)
+            }
+
+            // If we keep failing then the block must be almost full so bail out here.
+            if (nPackageFailures >= 5)
```
Let's make 5 a symbolic constant, I can imagine that we might want to tune this.

ptschip replied:
done

ptschip commented:
ok done

On 04/09/2019 6:05 p.m., Andrew Stone wrote:
>
> *@gandrewstone* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/globals.cpp
> <https://github.com/ptschip/BitcoinUnlimited/pull/6#discussion_r321037355>:
>
> > @@ -274,6 +275,9 @@ CTweakRef<bool> displayArchInSubver("net.displayArchInSubver",
>      "Show box architecture, 32/64bit, in node user agent string (subver)",
>      &fDisplayArchInSubver);
>  
> +CTweak<bool> miningCPFP("mining.cpfp",
>
> let's spell out mining.cpfp as mining.childPaysForParent
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ptschip/BitcoinUnlimited/pull/6?email_source=notifications&email_token=ADCELUHTXTGEBETZO7YWTETQIBLNLA5CNFSM4ITMZKO2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCDWRNKY#pullrequestreview-283973291>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADCELUGA652WJR7GTNYGC2TQIBLNLANCNFSM4ITMZKOQ>.
>

ptschip commented:
fixed spelling

as for being part of block assembler, i don't think it makes
sense...these are global vars used to track peformance in the
miner_tests.cpp peformancetest section.

On 04/09/2019 6:08 p.m., Andrew Stone wrote:
>
> *@gandrewstone* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/miner.cpp
> <https://github.com/ptschip/BitcoinUnlimited/pull/6#discussion_r321037798>:
>
> >  #include <boost/tuple/tuple.hpp>
>  #include <queue>
>  #include <thread>
>  
> +// Track timing information for Score and Pacakge mining.
>
> spelling: Package
>
> Also, I'm just going from top to bottom here so may be missing
> something but shouldn't these be members of BlockAssembler?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ptschip/BitcoinUnlimited/pull/6?email_source=notifications&email_token=ADCELUA45IXRBXUXWJZ6AYLQIBLYLA5CNFSM4ITMZKO2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCDWRRTA#pullrequestreview-283973836>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADCELUB4P7G7Z5XVZKLBDX3QIBLYLANCNFSM4ITMZKOQ>.
>

ptschip commented:
done

On 04/09/2019 6:10 p.m., Andrew Stone wrote:
>
> *@gandrewstone* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/miner.h
> <https://github.com/ptschip/BitcoinUnlimited/pull/6#discussion_r321038235>:
>
> >  struct CBlockTemplate
>  {
>      CBlock block;
>      std::vector<CAmount> vTxFees;
>      std::vector<int64_t> vTxSigOps;
>  };
>  
> +
> +/** Comparator for CTxMemPool::txiter objects.
> + *  It simply compares the internal memory address of the CTxMemPoolEntry object
> + *  pointed to. This means it has no meaning, and is only useful for using them
> + *  as key in other indexes.
> + */
> +struct CompareCTxMemPoolIter
> +{
> +    bool operator()(const CTxMemPool::txiter &a, const CTxMemPool::txiter &b) const { return &(*a) < &(*b); }
> +};
> +
> +// A comparator that sorts transactions based on number of ancestors.
>
> use doxygen compatible comment format /**
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ptschip/BitcoinUnlimited/pull/6?email_source=notifications&email_token=ADCELUAVRAKIL7ZK47UYGEDQIBMCDA5CNFSM4ITMZKO2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCDWRWHQ#pullrequestreview-283974430>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADCELUEPVP36STAPC43XNMLQIBMCDANCNFSM4ITMZKOQ>.
>

gandrewstone commented on src/miner.cpp:
```diff
@@ -473,6 +515,141 @@ void BlockAssembler::addScoreTxs(std::vector<const CTxMemPoolEntry *> *vtxe)
+                continue;
+        }
+
+        // Test the package
```
not a good comment, need to answer why, and try to use different terms than the function name. Maybe "Check the block to determine if this package of transactions will fit within size and sigops limits".

ptschip replied:
done

ptschip commented:
done

On 04/09/2019 6:16 p.m., Andrew Stone wrote:
>
> *@gandrewstone* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/miner.cpp
> <https://github.com/ptschip/BitcoinUnlimited/pull/6#discussion_r321039254>:
>
> > +        if (packageFees < ::minRelayTxFee.GetFee(packageSize) && nBlockSize >= nBlockMinSize)
> +        {
> +            // Everything else we might consider has a lower fee rate so no need to continue
> +            return;
> +        }
> +
> +        // Test if package fits in the block
> +        if (nBlockSize + packageSize > nBlockMaxSize)
> +        {
> +            if (nBlockSize > nBlockMaxSize * .50)
> +            {
> +                nPackageFailures++;
> +            }
> +
> +            // If we keep failing then the block must be almost full so bail out here.
> +            if (nPackageFailures >= 5)
>
> Let's make 5 a symbolic constant, I can imagine that we might want to
> tune this.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ptschip/BitcoinUnlimited/pull/6?email_source=notifications&email_token=ADCELUBE4RNUOJMTE6WQOWTQIBMYFA5CNFSM4ITMZKO2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCDWR7QI#pullrequestreview-283975617>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADCELUD5GMIICS37GSE7I7TQIBMYFANCNFSM4ITMZKOQ>.
>

ptschip commented:
fixed, changed this to TestPackageSigOps()

On 04/09/2019 6:20 p.m., Andrew Stone wrote:
>
> *@gandrewstone* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/miner.cpp
> <https://github.com/ptschip/BitcoinUnlimited/pull/6#discussion_r321039904>:
>
> > +        // Test if package fits in the block
> +        if (nBlockSize + packageSize > nBlockMaxSize)
> +        {
> +            if (nBlockSize > nBlockMaxSize * .50)
> +            {
> +                nPackageFailures++;
> +            }
> +
> +            // If we keep failing then the block must be almost full so bail out here.
> +            if (nPackageFailures >= 5)
> +                return;
> +            else
> +                continue;
> +        }
> +
> +        // Test the package
>
> not a good comment, need to answer why, and try to use different terms
> than the function name. Maybe "Check the block to determine if this
> package of transactions will fit within size and sigops limits".
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ptschip/BitcoinUnlimited/pull/6?email_source=notifications&email_token=ADCELUDXBG5IZ3HINJJKP4LQIBNHHA5CNFSM4ITMZKO2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCDWSFUY#pullrequestreview-283976403>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADCELUB6J6UVTTJKRH2GNSDQIBNHHANCNFSM4ITMZKOQ>.
>

gandrewstone commented on src/miner.cpp:
```diff
@@ -473,6 +515,141 @@ void BlockAssembler::addScoreTxs(std::vector<const CTxMemPoolEntry *> *vtxe)
+// on feerate of a transaction including all unconfirmed ancestors.
+//
+// This is accomplished by considering a group of ancestors as a single transaction. We can call these
+// transations, Ancestor Grouped Transactions (AGT). This approach to grouping allows us to process
```
transations -> transactions

ptschip replied:
done

gandrewstone commented on src/miner.cpp:
```diff
@@ -473,6 +515,141 @@ void BlockAssembler::addScoreTxs(std::vector<const CTxMemPoolEntry *> *vtxe)
+// packages orders of magnitude faster than other methods of package mining since we no longer have
+// to continuously update the descendant state as we mine part of an unconfirmed chain.
+//
+// There is a theorical flaw in this approach which could happen when a block is almost full. We
```
theorical -> theoretical

ptschip replied:
done

gandrewstone commented on src/test/miner_tests.cpp:
```diff
@@ -81,6 +83,297 @@ bool TestSequenceLocks(const CTransaction &tx, int flags)
+    BOOST_CHECK(pblocktemplate->block.vtx[8]->GetHash() == hashLowFeeTx2);
+
+    // Test CPFP with AGT (ancestor grouped transactions)
+    // Add another 0 fee tx to higher fe tx chain. This should also get mined
```
fe -> fee

ptschip replied:
done

gandrewstone commented on src/test/miner_tests.cpp:
```diff
@@ -81,6 +83,297 @@ bool TestSequenceLocks(const CTransaction &tx, int flags)
+
+    // Test CPFP with AGT (ancestor grouped transactions)
+    // Add another 0 fee tx to higher fe tx chain. This should also get mined
+    // because the total package fees will still be above the minrelaytxfe
```
minrelaytxfe -> minrelaytxfee

ptschip replied:
done

gandrewstone commented on src/main.h:
```diff
@@ -75,13 +75,13 @@ static const CAmount HIGH_MAX_TX_FEE = 100 * HIGH_TX_FEE_PER_KB;
 static const unsigned int DEFAULT_MAX_ORPHAN_TRANSACTIONS = 1000000;
 /** Default for -limitancestorcount, max number of in-mempool ancestors */
-static const unsigned int DEFAULT_ANCESTOR_LIMIT = 25;
+static const unsigned int DEFAULT_ANCESTOR_LIMIT = 500;
```
I think we need to keep these at the BCH default, and then in a separate PR we need to implement XVERSION communication of the setting and preemptive forwarding of tx that would have been rejected from other nodes.

ptschip replied:
reverted the commit

gandrewstone commented on src/miner.cpp:
```diff
@@ -30,10 +30,19 @@
 #include <thread>
 
+// Track timing information for Score and Package mining.
+int64_t nTotalPackage = 0;
```
comment what critsec protects these globals... 
EDIT: actually I've looked closely at these now and they are only used in miner_tests GenerateBlocks which is single threaded.  So they could be garbage in the multi-threaded code and we would not care... but block creation is covered by cs_main anyway -- we only want to make 1 at a time.  So cs_main is protecting them

ptschip replied:
i made them atomic anyway

dagurval commented on src/miner.cpp:
```diff
@@ -473,6 +518,137 @@ void BlockAssembler::addScoreTxs(std::vector<const CTxMemPoolEntry *> *vtxe)
+    assert(it != mempool.mapTx.end());
+    if (inBlock.count(it))
+        return true;
+    return false;
```
nit: could just do `return inBlock.count(it)`

ptschip replied:
done, don't need this function anymore.

dagurval commented on src/miner.cpp:
```diff
@@ -473,6 +518,137 @@ void BlockAssembler::addScoreTxs(std::vector<const CTxMemPoolEntry *> *vtxe)
+    return false;
+}
+
+void BlockAssembler::SortForBlock(const CTxMemPool::setEntries &package, std::vector<CTxMemPool::txiter> &sortedEntries)
```
TTOR doesn't exist any more and I don't think the extra code complexity, and the extra test requirements for supporting it is worth it. Thus I think this method should be removed.

If TTOR support is really needed for some reason, then it can be added as a final sorting step (its not used in production, so  the performance hit of sorting is OK)

ptschip replied:
Same comment as below:  "Good point but we can't because we still have both in our regression tests such as ctor.py and also in the miner we later either do the sort or not based on the "canonical" flag. I think if we wan't to remove support for TTOR then it should be done in a different PR."

gandrewstone replied:
I agree, let's leave TTOR around for a while longer.

dagurval commented on src/miner.cpp:
```diff
@@ -473,6 +518,137 @@ void BlockAssembler::addScoreTxs(std::vector<const CTxMemPoolEntry *> *vtxe)
 }
 
+// Skip entries in mapTx that are already in a block.
+bool BlockAssembler::SkipMapTxEntry(CTxMemPool::txiter it)
```
nit: This method doesn't skip, maybe rename to `IsTxInBlock` or `ShouldSkipTx` ?

ptschip replied:
actually i just took this out and is not needed anymore..this was a leftover from Core's work where they did multiple checks here...i just replace skiptmapentry() with inBlock.count(iter)

dagurval commented on src/miner.cpp:
```diff
@@ -473,6 +518,137 @@ void BlockAssembler::addScoreTxs(std::vector<const CTxMemPoolEntry *> *vtxe)
+// this situation is not likely to be seen for two reasons. One, long unconfirmed chains are typically
+// having transactions with all the same fees and Two, the typical child pays for parent scenario has only
+// two transactions with the child having the higher fee. And neither of these two types of packages could
+// cause any loss of fees with this mining algorithm, when the block is nearly full.
```
Fair trade-off. More value in good performance over catching every satoshi.

ptschip replied:
yes

dagurval commented on src/miner.cpp:
```diff
@@ -473,6 +518,137 @@ void BlockAssembler::addScoreTxs(std::vector<const CTxMemPoolEntry *> *vtxe)
+            continue;
+        }
+
+        // The Package can now be added to the block.
```
We should just assume canonical (see comment above)

ptschip replied:
Good point but we can't because we still have both in our regression tests such as ctor.py and also in the miner we later either do the sort or not based on the "canonical" flag.  I think if we wan't to remove support for TTOR then it should be done in a different PR.

dagurval commented on src/txmempool.h:
```diff
@@ -298,7 +298,7 @@ class CompareTxMemPoolEntryByAncestorFee
         if (f1 == f2)
         {
-            return a.GetTx().GetHash() < b.GetTx().GetHash();
+            return a.GetTime() < b.GetTime();
```
Why? Is this related to CPFP?

ptschip replied:
Yes, gives a perfect fill in the block when for instance we have a long package which is greater than the maxblocksize (or the remaining amount of block to be filled) and all the fees for each transaction are equal.  This sorting does two things.  This way we end up finding the part of the package that first perfectly into the block, because we don't bail out early due to pacakge insertion failures, while also preserving some sense of fairness that the first transaction that arrived in the mempool has priority over the one that followed.  (Remember in BU and also ABC there is no longer any random delay in forwarding inv's anymore.)

dagurval commented on src/miner.cpp:
```diff
@@ -256,6 +256,30 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock(const CScript &sc
+        std::vector<const CTxMemPoolEntry *> vtxe;
+        addPriorityTxs(&vtxe);
+
+        // Mine by package (CPFP) or by score. Score mining is done following this so that we can fill any
```
> Score mining is done following this

Is it? Where?

ptschip replied:
thanks, was incorrect and an older comment.  Removed the last part.

dagurval commented:
Is it possible that ancestor package is 3MB? What happens then if `maxblocksize` is 2MB?

ptschip commented:
> Is it possible that ancestor package is 3MB? What happens then if `maxblocksize` is 2MB?

Very good question. Partly answered above with the  time sorting of the txns but what if the txns all have different fees?  Then the time sorting doesn't work and we have a potential problem because we can easily get 5 package failures and end up bailing out of building the block and end up with an empty block!  The solution  here is not to start counting package failures until we have partially filled the block.  If you look further down where we test if the package fits the block, you'll see we don't start counting package failures until we're > 50% full.
